gr/include/gnuradio/wait_strategy.hpp

```diff
@@ -0,0 +1,373 @@
+        std::unique_lock uniqueLock(_gate);
+        _conditionVariable.notify_all();
+    }
+};
+static_assert(WaitStrategy<TimeoutBlockingWaitStrategy>);
+static_assert(hasSignalAllWhenBlocking<TimeoutBlockingWaitStrategy>);
+
+/**
+ * Yielding strategy that uses a Thread.Yield() for IEventProcessors waiting on a barrier after an initially spinning.
+ * This strategy is a good compromise between performance and CPU resource without incurring significant latency spikes.
+ */
+class YieldingWaitStrategy {
+    const std::size_t _spinTries = 100;
+
+public:
+    std::int64_t waitFor(const std::int64_t sequence, const Sequence & /*cursor*/, const std::vector<std::shared_ptr<Sequence>> &dependentSequences) const {
+        auto       counter    = _spinTries;
+        const auto waitMethod = [&]() {
+            // optional: barrier check alert
+
+            if (counter == 0) {
+                std::this_thread::yield();
+            } else {
+                --counter;
+            }
+            return counter;
+        };
+
+        std::int64_t availableSequence;
+        while ((availableSequence = detail::getMinimumSequence(dependentSequences)) < sequence) {
+            counter = waitMethod();
+        }
+
+        return availableSequence;
+    }
+};
+static_assert(WaitStrategy<YieldingWaitStrategy>);
+static_assert(!hasSignalAllWhenBlocking<YieldingWaitStrategy>);
+
+struct NoWaitStrategy {
+    std::int64_t waitFor(const std::int64_t sequence, const Sequence & /*cursor*/, const std::vector<std::shared_ptr<Sequence>> & /*dependentSequences*/) const {
+        // wait for nothing
+        return sequence;
+    }
+};
+static_assert(WaitStrategy<NoWaitStrategy>);
+static_assert(!hasSignalAllWhenBlocking<NoWaitStrategy>);
+
+
+/**
+ *
+ * SpinWait is meant to be used as a tool for waiting in situations where the thread is not allowed to block.
+ *
+ * In order to get the maximum performance, the implementation first spins for `YIELD_THRESHOLD` times, and then
+ * alternates between yielding, spinning and putting the thread to sleep, to allow other threads to be scheduled
+ * by the kernel to avoid potential CPU contention.
+ *
+ * The number of spins, yielding, and sleeping for either '0 ms' or '1 ms' is controlled by the NTTP constants
+ * @tparam YIELD_THRESHOLD
+ * @tparam SLEEP_0_EVERY_HOW_MANY_TIMES
+ * @tparam SLEEP_1_EVERY_HOW_MANY_TIMES
+ */
+template<std::int32_t YIELD_THRESHOLD = 10, std::int32_t SLEEP_0_EVERY_HOW_MANY_TIMES = 5, std::int32_t SLEEP_1_EVERY_HOW_MANY_TIMES = 20>
+class SpinWait {
+    using Clock         = std::conditional_t<std::chrono::high_resolution_clock::is_steady, std::chrono::high_resolution_clock, std::chrono::steady_clock>;
+    std::int32_t _count = 0;
+    static void  spinWaitInternal(std::int32_t iterationCount) noexcept {
+        for (auto i = 0; i < iterationCount; i++) {
+            yieldProcessor();
+        }
+    }
+    static void yieldProcessor() noexcept { asm volatile("rep\nnop"); }
```

Comment:
um, slightly confused, what is assembler doing in a scheduling algorithm?

Reply:
that's a guaranteed 'NOOP' operation that the compiler cannot remove/optimise away. This is needed for ultra-high update rate when sometimes it's enough to wait for one clock cycle to resolve contention on a shared resource (here: atomics).

Reply:
if we care about that, then the 60ns race condition window we've discussed on call is a *humongous* problem…

Reply:
plus, I mean, makes no sense to have this in code that's not going to be compiled for x86, so hm.